File: regions/system/library/swift/migration/Sources/K1s0Migration/Migration.swift
import Foundation

// MARK: - Protocol

public protocol MigrationRunner: Sendable {
    func runUp() async throws -> MigrationReport
    func runDown(steps: Int) async throws -> MigrationReport
    func status() async throws -> [MigrationStatus]
    func pending() async throws -> [PendingMigration]
}

// MARK: - Config

public struct MigrationConfig: Sendable {
    public let migrationsDir: URL
    public let databaseUrl: String
    public let tableName: String

    public init(
        migrationsDir: URL,
        databaseUrl: String,
        tableName: String = "_migrations"
    ) {
        self.migrationsDir = migrationsDir
        self.databaseUrl = databaseUrl
        self.tableName = tableName
    }
}

// MARK: - Report

public struct MigrationReport: Sendable {
    public let appliedCount: Int
    public let elapsed: Duration
    public let errors: [any Error & Sendable]

    public init(appliedCount: Int, elapsed: Duration, errors: [any Error & Sendable] = []) {
        self.appliedCount = appliedCount
        self.elapsed = elapsed
        self.errors = errors
    }
}

// MARK: - Status

public struct MigrationStatus: Sendable {
    public let version: String
    public let name: String
    public let appliedAt: Date?
    public let checksum: String

    public init(version: String, name: String, appliedAt: Date?, checksum: String) {
        self.version = version
        self.name = name
        self.appliedAt = appliedAt
        self.checksum = checksum
    }
}

// MARK: - PendingMigration

public struct PendingMigration: Sendable {
    public let version: String
    public let name: String

    public init(version: String, name: String) {
        self.version = version
        self.name = name
    }
}

// MARK: - Error

public enum MigrationError: Error, Sendable {
    case connectionFailed(underlying: any Error & Sendable)
    case migrationFailed(version: String, underlying: any Error & Sendable)
    case checksumMismatch(version: String, expected: String, actual: String)
    case directoryNotFound(path: String)
    case rollbackNotSupported(version: String)
}

// MARK: - Direction

public enum MigrationDirection: Sendable {
    case up
    case down
}

// MARK: - File Parser

public struct MigrationFileParser: Sendable {
    public struct ParsedFile: Sendable {
        public let version: String
        public let name: String
        public let direction: MigrationDirection
    }

    public static func parseFilename(_ filename: String) -> ParsedFile? {
        guard filename.hasSuffix(".sql") else { return nil }
        let stem = String(filename.dropLast(4))

        let direction: MigrationDirection
        let rest: String

        if stem.hasSuffix(".up") {
            direction = .up
            rest = String(stem.dropLast(3))
        } else if stem.hasSuffix(".down") {
            direction = .down
            rest = String(stem.dropLast(5))
        } else {
            return nil
        }

        guard let idx = rest.firstIndex(of: "_"),
              idx > rest.startIndex,
              rest.index(after: idx) < rest.endIndex else {
            return nil
        }

        let version = String(rest[rest.startIndex..<idx])
        let name = String(rest[rest.index(after: idx)...])

        return ParsedFile(version: version, name: name, direction: direction)
    }

    public static func computeChecksum(_ content: String) -> String {
        let data = Array(content.utf8)
        var hash = [UInt8](repeating: 0, count: 32)
        // Simple SHA-256 alternative using built-in approach
        // Use a basic hash for portability without CryptoKit
        var h0: UInt64 = 0x6a09e667bb67ae85
        var h1: UInt64 = 0x3c6ef372a54ff53a
        for byte in data {
            h0 = h0 &* 31 &+ UInt64(byte)
            h1 = h1 &* 37 &+ UInt64(byte)
        }
        return String(format: "%016x%016x", h0, h1)
    }
}

// MARK: - InMemory Runner

public actor InMemoryMigrationRunner: MigrationRunner {
    private struct Entry: Sendable {
        let version: String
        let name: String
        let content: String
    }

    private let config: MigrationConfig
    private let upMigrations: [Entry]
    private let downMigrations: [String: Entry]
    private var applied: [MigrationStatus] = []

    public init(
        config: MigrationConfig,
        ups: [(version: String, name: String, content: String)],
        downs: [(version: String, name: String, content: String)]
    ) {
        self.config = config
        self.upMigrations = ups
            .map { Entry(version: $0.version, name: $0.name, content: $0.content) }
            .sorted { $0.version < $1.version }
        self.downMigrations = Dictionary(
            uniqueKeysWithValues: downs.map {
                ($0.version, Entry(version: $0.version, name: $0.name, content: $0.content))
            }
        )
    }

    public func runUp() async throws -> MigrationReport {
        let start = ContinuousClock().now
        let appliedVersions = Set(applied.map(\.version))
        var count = 0

        for mf in upMigrations {
            if appliedVersions.contains(mf.version) { continue }
            let cs = MigrationFileParser.computeChecksum(mf.content)
            applied.append(MigrationStatus(
                version: mf.version,
                name: mf.name,
                appliedAt: Date(),
                checksum: cs
            ))
            count += 1
        }

        let elapsed = ContinuousClock().now - start
        return MigrationReport(appliedCount: count, elapsed: elapsed)
    }

    public func runDown(steps: Int) async throws -> MigrationReport {
        let start = ContinuousClock().now
        var count = 0

        for _ in 0..<steps {
            if applied.isEmpty { break }
            applied.removeLast()
            count += 1
        }

        let elapsed = ContinuousClock().now - start
        return MigrationReport(appliedCount: count, elapsed: elapsed)
    }

    public func status() async throws -> [MigrationStatus] {
        let appliedMap = Dictionary(uniqueKeysWithValues: applied.map { ($0.version, $0) })
        return upMigrations.map { mf in
            let cs = MigrationFileParser.computeChecksum(mf.content)
            let appliedStatus = appliedMap[mf.version]
            return MigrationStatus(
                version: mf.version,
                name: mf.name,
                appliedAt: appliedStatus?.appliedAt,
                checksum: cs
            )
        }
    }

    public func pending() async throws -> [PendingMigration] {
        let appliedVersions = Set(applied.map(\.version))
        return upMigrations
            .filter { !appliedVersions.contains($0.version) }
            .map { PendingMigration(version: $0.version, name: $0.name) }
    }
}

File: regions/system/library/go/migration/migration.go
package migration

import (
	"context"
	"crypto/sha256"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"sync"
	"time"
)

// MigrationRunner defines the interface for running database migrations.
type MigrationRunner interface {
	RunUp(ctx context.Context) (*MigrationReport, error)
	RunDown(ctx context.Context, steps int) (*MigrationReport, error)
	Status(ctx context.Context) ([]*MigrationStatus, error)
	Pending(ctx context.Context) ([]*PendingMigration, error)
}

// MigrationConfig holds the configuration for migration operations.
type MigrationConfig struct {
	MigrationsDir string
	DatabaseURL   string
	TableName     string
	Driver        string
}

// NewMigrationConfig creates a MigrationConfig with default values.
func NewMigrationConfig(migrationsDir, databaseURL string) MigrationConfig {
	return MigrationConfig{
		MigrationsDir: migrationsDir,
		DatabaseURL:   databaseURL,
		TableName:     "_migrations",
		Driver:        "postgres",
	}
}

// MigrationReport contains the result of a migration operation.
type MigrationReport struct {
	AppliedCount int
	Elapsed      time.Duration
	Errors       []error
}

// MigrationStatus represents the status of a single migration.
type MigrationStatus struct {
	Version   string
	Name      string
	AppliedAt *time.Time
	Checksum  string
}

// PendingMigration represents a migration that has not been applied.
type PendingMigration struct {
	Version string
	Name    string
}

// MigrationDirection represents up or down migration.
type MigrationDirection int

const (
	DirectionUp MigrationDirection = iota
	DirectionDown
)

type migrationFile struct {
	Version   string
	Name      string
	Direction MigrationDirection
	Content   string
}

// ParseFilename parses a migration filename into version, name, and direction.
func ParseFilename(filename string) (version, name string, direction MigrationDirection, ok bool) {
	if !strings.HasSuffix(filename, ".sql") {
		return "", "", 0, false
	}
	stem := strings.TrimSuffix(filename, ".sql")

	if strings.HasSuffix(stem, ".up") {
		direction = DirectionUp
		stem = strings.TrimSuffix(stem, ".up")
	} else if strings.HasSuffix(stem, ".down") {
		direction = DirectionDown
		stem = strings.TrimSuffix(stem, ".down")
	} else {
		return "", "", 0, false
	}

	idx := strings.Index(stem, "_")
	if idx <= 0 || idx >= len(stem)-1 {
		return "", "", 0, false
	}

	version = stem[:idx]
	name = stem[idx+1:]
	ok = true
	return
}

// Checksum computes a SHA-256 checksum of the given content.
func Checksum(content string) string {
	h := sha256.Sum256([]byte(content))
	return fmt.Sprintf("%x", h)
}

// InMemoryMigrationRunner implements MigrationRunner using in-memory state.
type InMemoryMigrationRunner struct {
	config         MigrationConfig
	upMigrations   []migrationFile
	downMigrations map[string]migrationFile
	applied        []*MigrationStatus
	mu             sync.Mutex
}

// NewInMemoryRunner creates an InMemoryMigrationRunner that reads migration files from disk.
func NewInMemoryRunner(cfg MigrationConfig) (*InMemoryMigrationRunner, error) {
	info, err := os.Stat(cfg.MigrationsDir)
	if err != nil || !info.IsDir() {
		return nil, fmt.Errorf("directory not found: %s", cfg.MigrationsDir)
	}

	entries, err := os.ReadDir(cfg.MigrationsDir)
	if err != nil {
		return nil, fmt.Errorf("failed to read directory: %w", err)
	}

	var ups []migrationFile
	downs := make(map[string]migrationFile)

	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		version, name, direction, ok := ParseFilename(entry.Name())
		if !ok {
			continue
		}
		content, err := os.ReadFile(filepath.Join(cfg.MigrationsDir, entry.Name()))
		if err != nil {
			return nil, fmt.Errorf("failed to read file %s: %w", entry.Name(), err)
		}
		mf := migrationFile{
			Version:   version,
			Name:      name,
			Direction: direction,
			Content:   string(content),
		}
		switch direction {
		case DirectionUp:
			ups = append(ups, mf)
		case DirectionDown:
			downs[version] = mf
		}
	}

	sort.Slice(ups, func(i, j int) bool {
		return ups[i].Version < ups[j].Version
	})

	return &InMemoryMigrationRunner{
		config:         cfg,
		upMigrations:   ups,
		downMigrations: downs,
		applied:        nil,
	}, nil
}

// NewInMemoryRunnerFromMigrations creates an InMemoryMigrationRunner from provided migration data.
func NewInMemoryRunnerFromMigrations(cfg MigrationConfig, ups []struct{ Version, Name, Content string }, downs []struct{ Version, Name, Content string }) *InMemoryMigrationRunner {
	var upFiles []migrationFile
	downFiles := make(map[string]migrationFile)

	for _, u := range ups {
		upFiles = append(upFiles, migrationFile{
			Version:   u.Version,
			Name:      u.Name,
			Direction: DirectionUp,
			Content:   u.Content,
		})
	}
	for _, d := range downs {
		downFiles[d.Version] = migrationFile{
			Version:   d.Version,
			Name:      d.Name,
			Direction: DirectionDown,
			Content:   d.Content,
		}
	}

	sort.Slice(upFiles, func(i, j int) bool {
		return upFiles[i].Version < upFiles[j].Version
	})

	return &InMemoryMigrationRunner{
		config:         cfg,
		upMigrations:   upFiles,
		downMigrations: downFiles,
		applied:        nil,
	}
}

func (r *InMemoryMigrationRunner) RunUp(_ context.Context) (*MigrationReport, error) {
	r.mu.Lock()
	defer r.mu.Unlock()

	start := time.Now()
	appliedSet := make(map[string]bool)
	for _, a := range r.applied {
		appliedSet[a.Version] = true
	}

	count := 0
	for _, mf := range r.upMigrations {
		if appliedSet[mf.Version] {
			continue
		}
		now := time.Now().UTC()
		r.applied = append(r.applied, &MigrationStatus{
			Version:   mf.Version,
			Name:      mf.Name,
			AppliedAt: &now,
			Checksum:  Checksum(mf.Content),
		})
		count++
	}

	return &MigrationReport{
		AppliedCount: count,
		Elapsed:      time.Since(start),
	}, nil
}

func (r *InMemoryMigrationRunner) RunDown(_ context.Context, steps int) (*MigrationReport, error) {
	r.mu.Lock()
	defer r.mu.Unlock()

	start := time.Now()
	count := 0

	for i := 0; i < steps; i++ {
		if len(r.applied) == 0 {
			break
		}
		r.applied = r.applied[:len(r.applied)-1]
		count++
	}

	return &MigrationReport{
		AppliedCount: count,
		Elapsed:      time.Since(start),
	}, nil
}

func (r *InMemoryMigrationRunner) Status(_ context.Context) ([]*MigrationStatus, error) {
	r.mu.Lock()
	defer r.mu.Unlock()

	appliedMap := make(map[string]*MigrationStatus)
	for _, a := range r.applied {
		appliedMap[a.Version] = a
	}

	var statuses []*MigrationStatus
	for _, mf := range r.upMigrations {
		checksum := Checksum(mf.Content)
		if applied, ok := appliedMap[mf.Version]; ok {
			statuses = append(statuses, &MigrationStatus{
				Version:   mf.Version,
				Name:      mf.Name,
				AppliedAt: applied.AppliedAt,
				Checksum:  checksum,
			})
		} else {
			statuses = append(statuses, &MigrationStatus{
				Version:  mf.Version,
				Name:     mf.Name,
				Checksum: checksum,
			})
		}
	}

	return statuses, nil
}

func (r *InMemoryMigrationRunner) Pending(_ context.Context) ([]*PendingMigration, error) {
	r.mu.Lock()
	defer r.mu.Unlock()

	appliedSet := make(map[string]bool)
	for _, a := range r.applied {
		appliedSet[a.Version] = true
	}

	var pending []*PendingMigration
	for _, mf := range r.upMigrations {
		if !appliedSet[mf.Version] {
			pending = append(pending, &PendingMigration{
				Version: mf.Version,
				Name:    mf.Name,
			})
		}
	}

	return pending, nil
}
